axelor-budget/src/main/java/com/axelor/apps/budget/service/BudgetToolsServiceImpl.java

```diff
@@ -72,4 +82,56 @@ public boolean checkBudgetKeyAndRoleForMove(Move move) throws AxelorException {
     }
     return false;
   }
+
+  @Override
+  public GlobalBudget getGlobalBudgetUsingBudget(Budget budget) {
+
+    if (budget == null) {
+      return null;
+    }
+
+    if (budget.getGlobalBudget() != null) {
+      return budget.getGlobalBudget();
+    }
+    if (budget.getBudgetLevel() != null) {
+      return getGlobalBudgetUsingBudgetLevel(budget.getBudgetLevel());
+    }
+
+    return null;
+  }
+
+  @Override
+  public GlobalBudget getGlobalBudgetUsingBudgetLevel(BudgetLevel budgetLevel) {
+    if (budgetLevel == null) {
+      return null;
+    }
+
+    if (budgetLevel.getGlobalBudget() != null) {
+      return budgetLevel.getGlobalBudget();
+    }
+
+    return getGlobalBudgetUsingBudgetLevel(budgetLevel.getParentBudgetLevel());
+  }
+
+  @Override
+  public String getBudgetExceedMessage(String budgetExceedAlert, boolean isOrder, boolean isError) {
+    if (Strings.isNullOrEmpty(budgetExceedAlert)) {
+      return "";
+    }
+    if (isOrder && isError) {
+      return budgetExceedAlert;
+    }
+    return String.format(
+        "%s %s",
+        I18n.get(budgetExceedAlert), I18n.get(BudgetExceptionMessage.BUDGET_EXCEED_ERROR_ALERT));
+  }
+
+  @Override
+  public boolean canAutoComputeBudgetDistribution(Company company, List<Object> list)
```

Comment:
you can replace `List<Object>` by `List<? extends Model>`
explanation here: https://docs.oracle.com/javase/tutorial/java/generics/upperBounded.html 

This way you can directly use the List of Invoice/Sale Order/... in the calling methods instead of `Collections.singletonList()`